File: raspberry-pi-ios-client/RaspberryPiClient/Data/Infrastructure/Foxglove/Server/FoxgloveAdvertise.swift
//
//  FoxgloveAdvertise.swift
//  RaspberryPiClient
//
//  Created by Shusuke Ota on 2025/03/20.
//

import Foundation

/// Advertise メッセージ（サーバー → クライアント）
/// 利用可能なチャンネル（トピック）をクライアントに通知する
/// Ref: https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#advertise
struct FoxgloveAdvertise: Codable {
    let op: FoxgloveMessageOperation
    let channels: [FoxgloveAdvertiseChannel]

    init(op: FoxgloveMessageOperation = .advertise, channels: [FoxgloveAdvertiseChannel]) {
        self.op = op
        self.channels = channels
    }

    /// トピック名 → channelId のマッピングを返す
    func topicToChannelIdMap() -> [String: UInt32] {
        Dictionary(uniqueKeysWithValues: channels.map { ($0.topic, $0.id) })
    }
}

/// Advertise メッセージ内のチャンネル情報（サーバー → クライアント）
/// Ref: https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#advertise
struct FoxgloveAdvertiseChannel: Codable {
    let id: UInt32
    let topic: String
    let encoding: String
    let schemaName: String
    let schema: String
    let schemaEncoding: String?
}

File: raspberry-pi-android-client/app/src/main/java/com/shusuke/raspberry_pi_client/data/infrastructure/foxglove/server/FoxgloveAdvertise.kt
package com.shusuke.raspberry_pi_client.data.infrastructure.foxglove.server

import com.shusuke.raspberry_pi_client.data.infrastructure.foxglove.common.FoxgloveMessageOperation
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Advertise メッセージ（サーバー → クライアント）
 *
 * 利用可能なチャンネル（トピック）をクライアントに通知する。
 * Ref: https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#advertise
 */
@Serializable
data class FoxgloveAdvertise(
    val op: FoxgloveMessageOperation = FoxgloveMessageOperation.Advertise,
    val channels: List<FoxgloveAdvertiseChannel> = emptyList(),
) {
    fun topicToChannelIdMap(): Map<String, UInt> =
        channels.associate { it.topic to it.id }

    fun topicToEncodingMap(): Map<String, String> =
        channels.associate { it.topic to it.encoding }
}

/**
 * Advertise メッセージ内のチャンネル情報（サーバー → クライアント）
 *
 * Ref: https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#advertise
 */
@Serializable
data class FoxgloveAdvertiseChannel(
    val id: UInt,
    val topic: String,
    val encoding: String,
    @SerialName("schemaName")
    val schemaName: String,
    val schema: String,
    @SerialName("schemaEncoding")
    val schemaEncoding: String? = null,
)
